WooCommerce/Classes/ViewModels/Refunds/RefundedProductsViewModel.swift

```diff
@@ -9,14 +9,88 @@ final class RefundedProductsViewModel {
     ///
     private(set) var order: Order
 
-    /// Array of all refunded items.
+    /// Array of all refunded items from every refund.
     ///
     private(set) var items: [OrderItemRefund]
 
```

Comment:
This is where the aggregated data work happens 👇

Reply:
Thanks for the comments along with the code, it was helpful. If I understand the goals of the refund summary correctly, this is my approach with O(n) with n being the number of refund items:

As I remember, a variation is like a Product and has a unique Product ID, with the logic like in
https://github.com/woocommerce/woocommerce-ios/blob/develop/WooCommerce/Classes/ViewModels/Order%20Details/OrderDetailsViewModel.swift#L197 :
```
let productID = item.variationID == 0 ? item.productID : item.variationID
```

(Maybe this can be an extension of `OrderItemRefund` like `item.productOrVariationID`)

so we could probably use this as the identifier for each refund item for a Product or Product Variation.

then we can keep a dictionary mapping from the above `productOrVariationID` to a set or array of refund items:

```
let itemsByProductID = [Int64: [OrderItemRefund]]()
```

then we can populate `itemsByProductID` by iterating over `items` and then add each `item` to the corresponding value given the `productID` calculation.

after `itemsByProductID` is populated, then we can process each Product/Variation ID and summarize all the items into an `OrderItemRefundSummary`.

feel free to correct anything if I misunderstood any part!

Reply:
I found this part a bit hard to follow, since it has an imperative style and one needs to keep track of what goes into each temporary array.

The way I understand the problem is that you have multiple refund items that you want to group, then show a summary/aggregate of each group. This is a common pattern that I'm used to see as SQL's `GROUP BY` and its aggregate functions (`COUNT`, `SUM`,...), so my background with that might make this proposal biased, but I feel like this would be clearer.

- First, group by the `productID`, `variationID` tuple, and you have one group for each row that you want to present
- Now each group is an array of items, so you need to aggregate their values somehow
- For the item details, take the first item (they should all be identical, right?), then sum the quantities and prices
- Generate an `OrderItemRefundSummary` with that data for each group.

```swift
/// Condense order items into summarized data
///
var summedItems: [OrderItemRefundSummary] {
    /// OrderItemRefund.orderItemID isn't useful for finding duplicates here,
    /// because multiple refunds cause orderItemIDs to be unique.
    /// Instead, we need to find duplicate *Products*.
    let currency = CurrencyFormatter()

    let grouped = Dictionary(grouping: items) { (item) in
        // This needs to be hashable so a tuple won't do it
        // I doubt this would become a performance issue, but it could be replaced
        // by a hashable struct
        return "\(item.productID)-\(item.variationID)"
    }
    return grouped.compactMap { (_, items) in
        // Here we iterate over each group's items

        // All items should be equal except for quantity and price, so we pick the first
        guard let item = items.first else {
            // This should never happen, but let's be safe
            return nil
        }

        // Sum the quantities
        let totalQuantity = items
            .map( { $0.quantity } )
            .reduce(0, +)
        // Sum the refunded amount
        let totalTax = items
            .compactMap( { currency.convertToDecimal(from: $0.totalTax) } )
            .reduce(NSDecimalNumber(value: 0), { $0.adding($1) })

        return OrderItemRefundSummary(
            name: item.name,
            productID: item.productID,
            variationID: item.variationID,
            quantity: totalQuantity,
            price: item.price,
            sku: item.sku,
            totalTax: totalTax)
    }
}
```

I feel like key paths could make the syntax a bit better, if we started adding a few helper extensions:

```swift
extension Sequence {
    func map<T>(_ keyPath: KeyPath<Element, T>) -> [T] {
        return map { $0[keyPath: keyPath] }
    }

    func sum<T: Numeric>(_ keyPath: KeyPath<Element, T>) -> T {
        return map(keyPath).sum()
    }
}

extension Sequence where Element: Numeric {
    /// Returns the sum of all elements in the collection
    func sum() -> Element { return reduce(0, +) }
}
```

Then the total quantity could be calculated as:

```swift
let totalQuantity = items.sum(\.quantity)
```

Reply:
Some quick notes, as I work my way through this 😀

>a variation is like a Product and has a unique Product ID

I think that is true for Products (a product that was pulled from the products results controller), but not true for Order Items (as I encountered that issue early when writing this PR). `OrderItem`s have unique productIDs, but the variationID is only unique within the context of a productID. So for example,  an `OrderItem` that was derived from a variable product `Blue Fish - SM` can have the same variationID (productID: 71, variationID: 22) as `Pink Penguin - L` (productID: 3, variationID: 22). 

> I found this part a bit hard to follow, since it has an imperative style and one needs to keep track of what goes into each temporary array.

Yes. I fell back on Objective-C thinking to create a working example. This is very ... _not_ Swift

Reply:
@jaclync thank you for proposing a better solution than what I had 🌟. I really appreciate it. @koke thank you for writing out an example of what you talked about before. I couldn't make the leap from what you explained to actually putting it into code.

I decided to go with @koke's solution. While researching how `Dictionary(grouping:by)` works, I learned that returning a hashable value creates an O(1) result (doesn't matter how many order items we're grouping)!

The code below is ready for another review.

Reply:
@koke's solution looks great and elegant 👍 TIL about the key paths mapping and `Dictionary(grouping:by:)`!

> a variation is like a Product and has a unique Product ID
> So for example, an OrderItem that was derived from a variable product Blue Fish - SM can have the same variationID (productID: 71, variationID: 22) as Pink Penguin - L (productID: 3, variationID: 22).

from my understanding, variation IDs and product IDs are unique per site, because they are all `Post`s behind the scene. Local product attributes are unique per product.

Reply:
@koke TIL on the `Dictionary(groping:by:)` as well, that one is VERY cool!!

@mindgraffiti apologies I'm late to the party!!! I would probably have gone in the lines of splitting the `[OrderItemRefund]` property into multiple smaller pieces (and maybe unit test that!).

**BUT** I think Koke's SQL-like approach is as elegant and readable as it gets. I think he really nailed it there!